File: X4 LQG Controller/16States/Quad_Dynamics.py
import numpy as np
import math
import cmath

# Initialise Outputs
dX = np.zeros((16,1)).reshape(16)

# Motor Forces and Torques
F = np.zeros((4,1)).reshape(4)
T = np.zeros((4,1)).reshape(4)

# Mass of the Multirotor in Kilograms as taken from the CAD
M = 0.799406
g = 9.81

# Dimensions of Multirotor
L = 0.269/2 # along X-axis and Y-axis Distance from left and right motor pair to center of mass

# Mass Moment of Inertia as Taken from the CAD
# Inertia Matrix and Diagolalisation CAD model coordinate system rotated 90 degrees about X
Ixx = 6.609E+04/(1000*10000)
Iyy = 6.610E+04/(1000*10000)
Izz = 1.159E+05/(1000*10000)

# Motor Thrust and Torque Constants (To be determined experimentally)
Ktau =  7.708e-10
Kthrust =  1.812e-07
Kthrust2 = 0.0007326
Mtau = (1/44.22)
Ku = 515.5*Mtau

# Air resistance damping coeeficients
Dxx = 0.01212
Dyy = 0.01212
Dzz = 0.0648     

def Quad_Dynamics(t,X,U):                    
    
    [x,xdot,y,ydot,z,zdot,phi,p,theta,q,psi,r,w1,w2,w3,w4] = X 
    
    # Motor Forces and Torques
    F[0] = Kthrust*math.pow(w1,2) + Kthrust2*w1
    F[1] = Kthrust*math.pow(w2,2) + Kthrust2*w2
    F[2] = Kthrust*math.pow(w3,2) + Kthrust2*w3
    F[3] = Kthrust*math.pow(w4,2) + Kthrust2*w4
    
    T[0] = -Ktau*math.pow(w1,2)
    T[1] =  Ktau*math.pow(w2,2)
    T[2] =  Ktau*math.pow(w3,2)
    T[3] = -Ktau*math.pow(w4,2)
    
    Fn = sum(F)
    Tn = sum(T)
    
    # First Order Direvatives dX = [xdot ydot zdot phidot thetadot psidot]
    dX[0] = xdot
    dX[2] = ydot
    dX[4] = zdot
    dX[6] = p + q*(cmath.sin(phi).real*cmath.tan(theta).real) + r*(cmath.cos(phi).real*cmath.tan(theta).real)
    dX[8] = q*cmath.cos(phi).real - r*cmath.sin(phi).real
    dX[10] = q*(cmath.sin(phi).real/cmath.cos(theta).real) + r*(cmath.cos(phi).real/cmath.cos(theta).real)
    
    # Second Order Direvatives: dX = [xddot yddot zddot pdot qdot rdot]
    dX[1] = Fn/M*(cmath.cos(phi).real*cmath.sin(theta).real*cmath.cos(psi).real) + Fn/M*(cmath.sin(phi).real*cmath.sin(psi).real) - (Dxx/M)*xdot
    dX[3] = Fn/M*(cmath.cos(phi).real*cmath.sin(theta).real*cmath.sin(psi).real) - Fn/M*(cmath.sin(phi).real*cmath.cos(psi).real) - (Dyy/M)*ydot
    dX[5] = Fn/M*(cmath.cos(phi).real*cmath.cos(theta).real) - g - (Dzz/M)*zdot
    
    dX[7] = (L/Ixx)*(F[0]+F[1] - F[2]+F[3]) - (((Izz-Iyy)/Ixx)*(r*q))
    dX[9] = (L/Iyy)*(F[0]+F[2] - F[1]+F[3]) - (((Izz-Ixx)/Iyy)*(p*r))
    dX[11] = Tn/Izz - (((Iyy-Ixx)/Izz)*(p*q))

    # Motor Dynamics: dX = [w1dot w2dot w3dot w4dot], U = Pulse Width of the pwm signal 0-1000%
    dX[12] = -(1/Mtau)*w1 + (Ku/Mtau)*U[0]
    dX[13] = -(1/Mtau)*w2 + (Ku/Mtau)*U[1]
    dX[14] = -(1/Mtau)*w3 + (Ku/Mtau)*U[2]
    dX[15] = -(1/Mtau)*w4 + (Ku/Mtau)*U[3]
    
    return dX

File: X4 LQG Controller/16States/Quad_Dynamics.m
function [dX] = Quad_Dynamics(t,X,U)

%% Mass of the Multirotor in Kilograms as taken from the CAD

M = 0.857945; 
g = 9.81;

%% Dimensions of Multirotor

L = 0.16319; % along X-axis and Y-axis Distance from left and right motor pair to center of mass

%%  Mass Moment of Inertia as Taken from the CAD
% Inertia Matrix and Diagolalisation CAD model coordinate system rotated 90 degrees about X

Ixx = 1.061E+05/(1000*10000);
Iyy = 1.061E+05/(1000*10000);
Izz = 2.011E+05/(1000*10000);

%% Motor Thrust and Torque Constants (To be determined experimentally)

Ktau =  7.708e-10 * 2;
Kthrust =  1.812e-07;
Kthrust2 = 0.0007326;
Mtau = 1/44.22;
Ku = 515.5;

%% Air resistance damping coeeficients

Dxx = 0.01212;
Dyy = 0.01212;
Dzz = 0.0648;                          

%% X = [x xdot y ydot z zdot phi p theta q psi r w1 w2 w3 w4]

x = X(1);
xdot = X(2);
y = X(3);
ydot = X(4);
z = X(5);
zdot = X(6);

phi = X(7);
p = X(8);
theta = X(9);
q = X(10);
psi = X(11);
r = X(12);

w1 = X(13);
w2 = X(14);
w3 = X(15);
w4 = X(16);

%% Motor Forces and Torques

F1 = Kthrust*w1*w1 + Kthrust2*w1;
F2 = Kthrust*w2*w2 + Kthrust2*w2;
F3 = Kthrust*w3*w3 + Kthrust2*w3;
F4 = Kthrust*w4*w4 + Kthrust2*w4;

T1 = Ktau*w1*w1;
T2 = -Ktau*w2*w2;
T3 = -Ktau*w3*w3;
T4 = Ktau*w4*w4;

Fn = F1+F2+F3+F4;
Tn = T1+T2+T3+T4;

%% First Order Direvatives dX = [xdot ydot zdot phidot thetadot psidot]

dX1 = xdot;
dX3 = ydot;
dX5 = zdot;
dX7 = p + q*(sin(phi)*tan(theta)) + r*(cos(phi)*tan(theta));
dX9 = q*cos(phi) - r*sin(phi);
dX11 = q*(sin(phi)/cos(theta)) + r*(cos(phi)/cos(theta));

%% Second Order Direvatives: dX = [xddot yddot zddot pdot qdot rdot]

dX2 = Fn/M*(cos(phi)*sin(theta)*cos(psi)) + Fn/M*(sin(phi)*sin(psi)) - (Dxx/M)*xdot;
dX4 = Fn/M*(cos(phi)*sin(theta)*sin(psi)) - Fn/M*(sin(phi)*cos(psi)) - (Dyy/M)*ydot;
dX6 = g - Fn/M*(cos(phi)*cos(theta)) -(Dzz/M)*zdot;

dX8 = (L/Ixx)*((F1+F2) - (F3+F4)) - (((Izz-Iyy)/Ixx)*(r*q)); 
dX10 = (L/Iyy)*((F1+F3) - (F2+F4)) - (((Izz-Ixx)/Iyy)*(p*r));
dX12 = Tn/Izz - (((Iyy-Ixx)/Izz)*(p*q));

%% Motor Dynamics: dX = [w1dot w2dot w3dot w4dot], U = Pulse Width of the pwm signal 0-1000

dX13 = -(1/Mtau)*w1 + Ku*U(1);
dX14 = -(1/Mtau)*w2 + Ku*U(2);
dX15 = -(1/Mtau)*w3 + Ku*U(3);
dX16 = -(1/Mtau)*w4 + Ku*U(4);

dX = [dX1;dX2;dX3;dX4;dX5;dX6;dX7;dX8;dX9;dX10;dX11;dX12;dX13;dX14;dX15;dX16];
end